lib/src/annotation_manager.dart

```diff
@@ -0,0 +1,334 @@
+    }
+  }
+
+  _onDrag(dynamic id,
+      {required Point<double> point,
+      required LatLng origin,
+      required LatLng current,
+      required LatLng delta}) {
+    final annotation = byId(id);
+    if (annotation != null) {
+      annotation.translate(delta);
+      set(annotation);
+    }
+  }
+
+  /// Set an existing anntotation to the map. Use this to do a fast update for a
+  /// single annotation
+  Future<void> set(T annoation) async {
+    assert(_idToAnnotation.containsKey(annoation.id),
+        "you can only set existing annotations");
+    _idToAnnotation[annoation.id] = annoation;
+    final oldLayerIndex = _idToLayerIndex[annoation.id];
+    final layerIndex = selectLayer != null ? selectLayer!(annoation) : 0;
+    if (oldLayerIndex != layerIndex) {
+      // if the annotation has to be moved to another layer/source we have to
+      // set all
+      await _setAll();
+    } else {
+      await controller.setGeoJsonFeature(
+          _makeLayerId(layerIndex), annoation.toGeoJson());
+    }
+  }
+}
+
+class LineManager extends AnnotationManager<Line> {
+  LineManager(MapboxMapController controller,
+      {void Function(Line)? onTap, bool enableInteraction = true})
+      : super(
+          controller,
+          onTap: onTap,
+          enableInteraction: enableInteraction,
+          selectLayer: (Line line) => line.options.linePattern == null ? 0 : 1,
+        );
+
+  static const _baseProperties = LineLayerProperties(
+    lineJoin: [Expressions.get, 'lineJoin'],
+    lineOpacity: [Expressions.get, 'lineOpacity'],
+    lineColor: [Expressions.get, 'lineColor'],
+    lineWidth: [Expressions.get, 'lineWidth'],
+    lineGapWidth: [Expressions.get, 'lineGapWidth'],
+    lineOffset: [Expressions.get, 'lineOffset'],
+    lineBlur: [Expressions.get, 'lineBlur'],
+  );
+  @override
+  List<LayerProperties> get allLayerProperties => [
+        _baseProperties,
+        _baseProperties.copyWith(
+            LineLayerProperties(linePattern: [Expressions.get, 'linePattern'])),
+      ];
+}
+
+class FillManager extends AnnotationManager<Fill> {
+  FillManager(
+    MapboxMapController controller, {
+    void Function(Fill)? onTap,
+    bool enableInteraction = true,
+  }) : super(
+          controller,
+          onTap: onTap,
+          enableInteraction: enableInteraction,
+          selectLayer: (Fill fill) => fill.options.fillPattern == null ? 0 : 1,
+        );
+  @override
+  List<LayerProperties> get allLayerProperties => const [
+        FillLayerProperties(
+          fillOpacity: [Expressions.get, 'fillOpacity'],
+          fillColor: [Expressions.get, 'fillColor'],
+          fillOutlineColor: [Expressions.get, 'fillOutlineColor'],
+        ),
+        FillLayerProperties(
+          fillOpacity: [Expressions.get, 'fillOpacity'],
+          fillColor: [Expressions.get, 'fillColor'],
+          fillOutlineColor: [Expressions.get, 'fillOutlineColor'],
+          fillPattern: [Expressions.get, 'fillPattern'],
```

Comment:
Why add an additional `FillLayerProperties` with `fillPattern`?

Reply:
Fill Pattern are mutually exclusive to other fill settings. Meaning if you set a fill pattern to a layer, it will disable all normal (without a pattern) rendering of fills for this layers.

To support Fills with and without patterns in the same fill manager i added two different layers, one with patterns and one without. The Annotation manager than calls selectLayer to select the proper layer for an annotation.

The same behavior is true for the LineManager.

With the current system Fills will stop rendering as soon as you add one with a pattern.

Reply:
The code for this can be simplified in the future, as soon as we support filters for layers. Which i feel should be one of next features to implement.

Reply:
I see, thanks.